src/V3EmitCHeaders.cpp

```diff
@@ -254,8 +263,115 @@ class EmitCHeader final : public EmitCConstInit {
+        } else if (auto* adtypep = VN_CAST(dtypep, NodeUOrStructDType)) {
+            std::string tmp = m_names.get("__tmp");
+            std::string suffixName = dtypep->isWide() ? tmp + ".data()" : "&" + tmp;
+            if (getfunc) {  // Emit `get` func;
+                // auto __tmp = field.get();
+                puts("auto " + tmp + " = " + fieldname + ".get();\n");
+                // bitHelper(&ret, offset, &__tmp, 0, width);
+                puts("bitHelper(" + retOrArg + ", " + offset + ", " + suffixName + ", 0, "
+                     + std::to_string(adtypep->width()) + ");\n");
+            } else {  // Emit `set` func
+                std::string tmptype = AstCDType::typeToHold(adtypep->width());
+                // type tmp;
+                puts(tmptype + " " + tmp + ";\n");
+                // bitHelper(&__tmp, 0, &retOrArg, offset, width);
+                puts("bitHelper(" + suffixName + ", 0, " + retOrArg + ", " + offset + ", "
+                     + std::to_string(adtypep->width()) + ");\n");
+                // field.set(__tmp);
+                puts(fieldname + ".set(" + tmp + ");\n");
+            }
+        } else {
+            UASSERT_OBJ(VN_IS(dtypep, EnumDType) || VN_IS(dtypep, BasicDType), dtypep,
+                        "Unsupported type in packed struct or union");
+            std::string suffixName = dtypep->isWide() ? fieldname + ".data()" : "&" + fieldname;
+            if (getfunc) {  // Emit `get` func;
+                // bitHelper(&ret, offset, &field, 0, width);
+                puts("bitHelper(" + retOrArg + ", " + offset + ", " + suffixName + ", 0, "
+                     + std::to_string(dtypep->width()) + ");\n");
+            } else {  // Emit `set` func
+                // bitHelper(&field, 0, &retOrArg, offset, width);
+                puts("bitHelper(" + suffixName + ", 0, " + retOrArg + ", " + offset + ", "
+                     + std::to_string(dtypep->width()) + ");\n");
+            }
+        }
+    }
+    void emitPackedUOrSBody(AstNodeUOrStructDType* sdtypep) {
+        putns(sdtypep, sdtypep->verilogKwd());  // "struct"/"union"
+        puts(" " + EmitCBase::prefixNameProtect(sdtypep) + " {\n");
+
+        AstMemberDType *itemp, *lastItemp;
+        AstMemberDType* witemp = nullptr;
+        // LSB is first field in C, so loop backwards
+        for (lastItemp = sdtypep->membersp(); lastItemp && lastItemp->nextp();
+             lastItemp = VN_AS(lastItemp->nextp(), MemberDType)) {
+            if (lastItemp->width() == sdtypep->width()) witemp = lastItemp;
+        }
+        for (itemp = lastItemp; itemp; itemp = VN_CAST(itemp->backp(), MemberDType)) {
+            putns(itemp, itemp->dtypep()->cType(itemp->nameProtect(), false, false, true));
+            puts(";\n");
+        }
+        // Emit constructor to zero all bytes
+        puts(EmitCBase::prefixNameProtect(sdtypep)
+             + "() { memset(reinterpret_cast<void*>(this), 0, sizeof(*this)); }\n");
```

Comment:
Why zero?  This will cause problems if we ever allow other non-binary types.

Nit: use std::memset

Reply:
`vlBitHelper` will not clear upper unused bits. So when users write something like
```c++
exported_packed_struct_t struct;
struct.set(top_module->outsignals)  // fill each field of struct with output signals
cout << struct.some_field;
```
if `some_field` is representing a 5 bits signal, upper three bits of `struct.some_field` will be garbage. Users don't need to clear these bits if we zero each field in constructor.

But maybe it's better to clear unused upper bits in `vlBitHelper`, so I add a `clean` argument to `vlBitHelper` to determine whether upper bits will be cleared. Now we don't emit this constructor.